File: tests/cpp/test_enkf.cpp
#include <gtest/gtest.h>

#include <cmath>
#include <vector>

#include "nephele/ensemble_kalman.hpp"

using namespace nephele;

// A scalar observation that directly measures the (single) state component must
// pull the ensemble mean toward the observation.
TEST(EnKF, ScalarUpdateMovesMeanTowardObservation) {
    EnsembleKalmanFilter enkf(1, 4000);

    std::vector<std::vector<double>> members;
    members.reserve(4000);
    // Forecast ensemble centred at 0 with spread 1.
    double phase = 0.0;
    for (std::size_t e = 0; e < 4000; ++e) {
        phase += 0.6180339887;  // low-discrepancy-ish deterministic spread
        const double g = std::sqrt(-2.0 * std::log((e + 1.0) / 4001.0)) *
                         std::cos(6.2831853 * std::fmod(phase, 1.0));
        members.push_back({g});
    }
    enkf.set_ensemble(members);

    const double prior_mean = enkf.mean()[0];

    // Observe y = 5 with H = [1], R = 0.25.
    enkf.update({1.0}, {5.0}, {0.25}, 42);

    const double post_mean = enkf.mean()[0];
    EXPECT_GT(post_mean, prior_mean);
    EXPECT_LT(post_mean, 5.0);  // shrinks toward but not past the observation
    EXPECT_GT(post_mean, 2.0);  // meaningful correction given R < prior var
}

TEST(EnKF, RejectsInvalidShapes) {
    EnsembleKalmanFilter enkf(2, 10);
    EXPECT_THROW(enkf.update({1.0}, {1.0, 2.0}, {0.1, 0.1}), std::invalid_argument);
    EXPECT_THROW(enkf.update({1.0, 0.0}, {1.0}, {-0.1}), std::invalid_argument);
}

TEST(EnKF, ConstructorValidatesDimensions) {
    EXPECT_THROW(EnsembleKalmanFilter(0, 10), std::invalid_argument);
    EXPECT_THROW(EnsembleKalmanFilter(2, 1), std::invalid_argument);
}

File: tests/python/test_enkf.py
import numpy as np
import pytest

from nephele.assimilation.enkf import EnsembleKalmanFilter


def test_enkf_scalar_update_moves_mean_toward_observation():
    rng = np.random.default_rng(0)
    ensemble = rng.normal(loc=0.0, scale=1.0, size=(2000, 1))
    enkf = EnsembleKalmanFilter(ensemble, seed=1)

    prior = enkf.mean()[0]
    enkf.update(np.array([[1.0]]), np.array([5.0]), np.array([0.25]))
    post = enkf.mean()[0]

    assert post > prior
    assert post < 5.0
    assert post > 2.0  # meaningful correction since R < prior variance


def test_enkf_multivariate_reduces_spread():
    rng = np.random.default_rng(2)
    ensemble = rng.normal(size=(3000, 2)) @ np.array([[1.0, 0.4], [0.0, 1.0]])
    enkf = EnsembleKalmanFilter(ensemble, seed=3)

    prior_var = enkf.ensemble.var(axis=0)
    h = np.array([[1.0, 0.0]])  # observe first component
    enkf.update(h, np.array([0.0]), np.array([0.1]))
    post_var = enkf.ensemble.var(axis=0)

    # Observed component variance must shrink.
    assert post_var[0] < prior_var[0]


def test_enkf_rejects_bad_shapes():
    enkf = EnsembleKalmanFilter(np.zeros((10, 2)))
    with pytest.raises(ValueError):
        enkf.update(np.zeros((1, 3)), np.zeros(1), np.ones(1))
    with pytest.raises(ValueError):
        enkf.update(np.zeros((1, 2)), np.zeros(1), np.array([-1.0]))


def test_enkf_rejects_small_ensemble():
    with pytest.raises(ValueError):
        EnsembleKalmanFilter(np.zeros((1, 2)))
